cpp/src/gis/cuda/common/gis_definitions.cpp

```diff
@@ -0,0 +1,189 @@
+  assert(value_offsets_[size] == values_.size());
+  holder->size = size;
+  holder->tags = GpuAlloc<WkbTag>(tags_.size());
+  holder->meta_offsets = GpuAlloc<int>(meta_offsets_.size());
+  holder->value_offsets = GpuAlloc<int>(value_offsets_.size());
+  holder->data_state = data_state_;
+  assert(holder->metas == nullptr);
+  assert(holder->values == nullptr);
+  return holder;
+}
+
+void GeometryVector::OutputEvolveWith(GpuContext& gpu_ctx) {
+  assert(data_state_ == DataState::FlatOffset_EmptyInfo);
+  assert(gpu_ctx.data_state == DataState::FlatOffset_FullInfo);
+  assert(tags_.size() == gpu_ctx.size);
+  auto size = gpu_ctx.size;
+
+  // copy tags to gpu_ctx
+  GpuMemcpy(tags_.data(), gpu_ctx.tags, size);
+
+  // exclusive scan offsets[0, n+1) in gpu_ctx, and copy to cpu
+  // return offsets[n] as total size
+  auto scan_at = [size](int* gpu_addr, int* cpu_addr) {
+    int zero = 0;
+    GpuMemcpy(gpu_addr + size, &zero, 1);
+    internal::ExclusiveScan(gpu_addr, size + 1);
+    GpuMemcpy(cpu_addr, gpu_addr, size + 1);
+    return cpu_addr[size];
+  };
+  auto meta_size = scan_at(gpu_ctx.meta_offsets, meta_offsets_.data());
+  auto value_size = scan_at(gpu_ctx.value_offsets, value_offsets_.data());
+
+  // alloc space for data(metas and values), for cpu and gpu.
+  metas_.resize(meta_size);
+  values_.resize(value_size);
+  data_state_ = DataState::PrefixSumOffset_EmptyData;
+  gpu_ctx.metas = GpuAlloc<uint32_t>(meta_size);
+  gpu_ctx.values = GpuAlloc<double>(value_size);
+  gpu_ctx.data_state = DataState::PrefixSumOffset_EmptyData;
+}
+
+void GeometryVector::OutputFinalizeWith(const GpuContext& gpu_ctx) {
+  assert(gpu_ctx.data_state == DataState::PrefixSumOffset_FullData);
+  assert(data_state_ == DataState::PrefixSumOffset_EmptyData);
+  assert(tags_.size() == gpu_ctx.size);
+  GpuMemcpy(metas_.data(), gpu_ctx.metas, metas_.size());
+  GpuMemcpy(values_.data(), gpu_ctx.values, values_.size());
+  data_state_ = DataState::PrefixSumOffset_FullData;
+}
+
+// only for testing
+// create Geometry from WktArray
+namespace GeometryVectorFactory {
+
+GeometryVector CreateFromWkts(const std::vector<std::string>& wkt_vec) {
+  auto input = WktsToArrowWkb(wkt_vec);
+  return ArrowWkbToGeometryVector(input);
+}
+
+GeometryVector CreateFromWkbs(const std::vector<std::vector<char>>& wkb_vec) {
+  arrow::BinaryBuilder builder;
+  for (const auto& wkb : wkb_vec) {
+    auto st = builder.Append(wkb.data(), wkb.size());
+    assert(st.ok());
+  }
+  std::shared_ptr<arrow::Array> arrow_wkb;
+  auto st = builder.Finish(&arrow_wkb);
+  assert(st.ok());
+  auto result = ArrowWkbToGeometryVector(arrow_wkb);
+  return result;
+}
+
+}  // namespace GeometryVectorFactory
+
+void GeometryVector::clear() {
+  tags_.clear();
+  metas_.clear();
+  values_.clear();
+  meta_offsets_.clear();
+  value_offsets_.clear();
+  data_state_ = DataState::Invalid;
+}
+
+// only for testing
+bool test_cuda_abi(const std::string& str) {
```

Comment:
move test_cuda_abi to test dir

Reply:
delete it since no use from the beginning.